zypp-plugin/snapper_zypp_plugin.cc

```diff
@@ -0,0 +1,341 @@
+	cerr << "INFO:" << "COMMITEND" << endl;
+
+	if (pre_snapshot_num != 0) {
+	    set<string> solvables = get_solvables(msg, Phase::AFTER);
+	    cerr << "DEBUG:" << "solvables: " << solvables << endl;
+
+	    bool found, important;
+	    match_solvables(solvables, found, important);
+	    cerr << "INFO:" << "found: " << found << ", important: " << important << endl;
+
+	    if (found || important) {
+		userdata["important"] = important ? "yes" : "no";
+
+		try {
+		    cerr << "INFO:" << "setting snapshot data" << endl;
+		    snapper::SMD modification_data;
+		    modification_data.description = snapshot_description;
+		    modification_data.cleanup = cleanup_algorithm;
+		    modification_data.userdata = userdata;
+		    command_set_snapshot(
+                        dbus_conn, snapper_cfg,
+                        pre_snapshot_num, modification_data
+                    );
+		}
+		catch (const DBus::ErrorException& ex) {
+		    SN_CAUGHT(ex);
+		    cerr << "ERROR:" << error_description(ex) << endl;
+		}
+		catch (const Exception& ex) {
+		    SN_CAUGHT(ex);
+		}
+		try {
+		    cerr << "INFO:" << "creating post snapshot" << endl;
+		    unsigned int post_snapshot_num = command_create_post_snapshot(
+                        dbus_conn, snapper_cfg,
+                        pre_snapshot_num, "", cleanup_algorithm, userdata
+                    );
+		    cerr << "DEBUG:" << "created post snapshot " << post_snapshot_num << endl;
+		}
+		catch (const DBus::ErrorException& ex) {
+		    SN_CAUGHT(ex);
+		    cerr << "ERROR:" << error_description(ex) << endl;
+		}
+		catch (const Exception& ex) {
+		    SN_CAUGHT(ex);
+		}
+	    }
+	    else {
+		try {
+		    cerr << "INFO:" << "deleting pre snapshot" << endl;
+		    vector<unsigned int> nums{ pre_snapshot_num };
+		    bool verbose = false;
+		    command_delete_snapshots(dbus_conn, snapper_cfg, nums, verbose);
+		    cerr << "DEBUG:" << "deleted pre snapshot " << pre_snapshot_num << endl;
+		}
+		catch (const DBus::ErrorException& ex) {
+		    SN_CAUGHT(ex);
+		    cerr << "ERROR:" << error_description(ex) << endl;
+		}
+		catch (const Exception& ex) {
+		    SN_CAUGHT(ex);
+		}
+	    }
+	}
+	return ack();
+    }
+
+private:
+    static const string cleanup_algorithm;
+
+    string snapper_cfg;
+
+    DBus::Connection dbus_conn;
+    unsigned int pre_snapshot_num;
+    string snapshot_description;
+    map<string, string> userdata;
+
+    vector<SolvableMatcher> solvable_matchers;
+
+    map<string, string> get_userdata(const Message&);
+
+    enum class Phase { BEFORE, AFTER };
+
+    set<string> get_solvables(const Message&, Phase phase);
+
+    void match_solvables(const set<string>& solvables, bool& found, bool& important);
+
+    unsigned int create_pre_snapshot(string config_name, string description, string cleanup, map<string, string> userdata);
+};
+
+const string SnapperZyppPlugin::cleanup_algorithm = "number";
+
+map<string, string> SnapperZyppPlugin::get_userdata(const Message& msg) {
```

Comment:
NP: why are there some methods defined outside the class declaration but others inside?

Reply:
Um, that is kind of random. Originally I put the "main logic" inline and the helpers outside.